lib/ts/recipe/multifactorauth/types.ts

```diff
@@ -0,0 +1,152 @@
+
+export type MFARequirementList = (
+    | {
+          oneOf: string[];
+      }
+    | {
+          allOfInAnyOrder: string[];
+      }
+    | string
+)[];
+
+export type MFAClaimValue = {
+    c: Record<string, number> | undefined;
+    v: boolean;
+};
+
+export type TypeInput = {
+    firstFactors?: string[];
+
+    override?: {
+        functions?: (
+            originalImplementation: RecipeInterface,
+            builder?: OverrideableBuilder<RecipeInterface>
+        ) => RecipeInterface;
+        apis?: (originalImplementation: APIInterface, builder?: OverrideableBuilder<APIInterface>) => APIInterface;
+    };
+};
+
+export type TypeNormalisedInput = {
+    firstFactors?: string[];
+
+    override: {
+        functions: (
+            originalImplementation: RecipeInterface,
+            builder?: OverrideableBuilder<RecipeInterface>
+        ) => RecipeInterface;
+        apis: (originalImplementation: APIInterface, builder?: OverrideableBuilder<APIInterface>) => APIInterface;
+    };
+};
+
+export type RecipeInterface = {
+    assertAllowedToSetupFactorElseThrowInvalidClaimError: (input: {
+        session: SessionContainer;
+        factorId: string;
+        mfaRequirementsForAuth: MFARequirementList;
+        factorsSetUpForUser: string[];
+        requiredSecondaryFactorsForUser: string[];
+        requiredSecondaryFactorsForTenant: string[];
+        completedFactors: Record<string, number>;
+        userContext: UserContext;
+    }) => Promise<void>;
+
+    getMFARequirementsForAuth: (input: {
+        user: User;
+        accessTokenPayload: JSONObject;
+        tenantId: string;
+        factorsSetUpForUser: string[];
+        requiredSecondaryFactorsForUser: string[];
+        requiredSecondaryFactorsForTenant: string[];
+        completedFactors: Record<string, number>;
+        userContext: UserContext;
+    }) => Promise<MFARequirementList> | MFARequirementList;
+
+    markFactorAsCompleteInSession: (input: {
+        session: SessionContainerInterface;
+        factorId: string;
+        userContext: UserContext;
+    }) => Promise<void>;
+
+    getFactorsSetupForUser: (input: { user: User; userContext: UserContext }) => Promise<string[]>;
+
+    getRequiredSecondaryFactorsForUser: (input: { userId: string; userContext: UserContext }) => Promise<string[]>;
+
+    addToRequiredSecondaryFactorsForUser: (input: {
+        userId: string;
+        factorId: string;
+        userContext: UserContext;
+    }) => Promise<void>;
+
+    removeFromRequiredSecondaryFactorsForUser: (input: {
+        userId: string;
+        factorId: string;
+        userContext: UserContext;
+    }) => Promise<void>;
+};
+
+export type APIOptions = {
+    recipeImplementation: RecipeInterface;
+    recipeInstance: Recipe;
+    config: TypeNormalisedInput;
+    recipeId: string;
+    isInServerlessEnv: boolean;
+    req: BaseRequest;
+    res: BaseResponse;
+};
+
+export type APIInterface = {
+    resyncSessionAndFetchMFAInfoPUT: (input: {
```

Comment:
api can also be undefined.. see APIInterface in other recipes!

Reply:
Done